### PR TITLE
Update to latest workflows templates

### PR DESCRIPTION
**Pull request by samuelallan72**
Also move tox.ini to repo root (makes it easier to work with) and remove sonar workflow (no longer used).

### CODE REVIEW

**Review by Pjack — DISMISSED** [review]

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -13,17 +13,26 @@ jobs:
 
   release:
     needs: check
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.runs-on }}
+    strategy:
+      fail-fast: false
+      matrix:
+        runs-on: ${runs_on}
```
Comment (chanchiwai-ray) [inline .github/workflows/release.yaml:20]: I think this is typo

Comment (samuelallan72) [reply .github/workflows/release.yaml:20]: Thanks, good catch!  Fixed :)

**Review by chanchiwai-ray — APPROVED** [review]

File: .github/workflows/promote.yaml [hunk 2]
```diff
@@ -1,25 +1,35 @@
...
     name: Promote charm
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
+      - name: Set channels
+        id: set-channels
+        run: |
+          channel_promotion="${{ github.event.inputs.channel-promotion }}"
+          origin=$(echo "$channel_promotion" | sed 's/\s*->.*//')
+          destination=$(echo "$channel_promotion" | sed 's/.*->\s*//')
```
Comment (aieri) [inline .github/workflows/promote.yaml:25]: suggestion: I think awk would be more readable here
```suggestion
          fs=' -> '
          origin=$(echo "$channel_promotion" | awk -F "$fs" '{print $1}')
          destination=$(echo "$channel_promotion" | awk -F "$fs" '{print $2}')
```

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -13,17 +13,27 @@ jobs:
...
+
       - name: Initialize lxd  # This should dropped once it's implemented on charming-actions itself. https://github.com/canonical/charming-actions/issues/140
-        uses: canonical/setup-lxd@v0.1.1
-      - name: Upload charm to charmhub
-        uses: canonical/charming-actions/upload-charm@2.4.0
+        # revision is latest main at time of writing; using because it contains a fix to https://github.com/canonical/setup-lxd/issues/19
+        uses: canonical/setup-lxd@2aa6f7caf7d1484298a64192f7f63a6684e648a4
+
+      - name: Pack and upload to charmhub
+        uses: canonical/charming-actions/upload-charm@2.6.2
```
Comment (aieri) [inline .github/workflows/release.yaml:33]: should we stick with 2.6.1 for now due to https://github.com/canonical/charming-actions/issues/148 ?

**Review by aieri — APPROVED** [review]
generally +1 with a couple of suggestions

File: .github/workflows/release.yaml (hunk 3 above)
Comment (samuelallan72) [reply .github/workflows/release.yaml:33]: We've been using 2.6.2 for several projects and haven't observed that bug. :)

File: .github/workflows/promote.yaml (hunk 2 above)
Comment (samuelallan72) [reply .github/workflows/promote.yaml:25]: Probably, but is it ok if we refactor later?  This simply copies our existing [template from the automation repo](https://github.com/canonical/solutions-engineering-automation/blob/2ee73f75536bb936a251c37817db4c1ba0892621/terraform-plans/templates/github/charm_promote.yaml.tftpl) - ideally we would do any refactorings from the automation repo so we can keep them all consistent across projects.

**Pull request merged**